3 -  CDC Proposal (#4)
gatewayd-io/proposals: 001-cdc/001---cdc.md

Advantages:
- Real-time event-driven capturing of data changes.
- Detection of all change event types: INSERTs, UPDATEs, and DELETES.
- Low impact on PostgreSQL performance.

Disadvantages:
- Not supported by very old PostgreSQL versions (older than 9.4).

Open source tools like [Debezium](https://github.com/debezium/debezium) use this feature, publishing changes to Kafka topics. Steps to use Debezium include configuring PostgreSQL, starting Kafka, starting Debezium, and implementing a Kafka consumer.

This image shows a high-level design of the system implemented in this approach.
![Pub Sub Design](./assets/pub-sub-design.svg)

For the subscriber/consumer part, we can try using the [Confluent PostgreSQL sink connector](https://docs.confluent.io/cloud/current/connectors/cc-postgresql-sink.html), and if there are any problems in the scale with this connector, we can consider replacing it with our in-house consumer in the future.

### Approach 3: Implement our publisher and subscriber in Go, using the PostgreSQL's logical replication

> In this approach, we will implement something similar to Debezium server/engine, but using Golang libraries such as [wal2json](https://github.com/eulerto/wal2json) and [wal-g](https://github.com/wal-g/wal-g).

Reviewer: ```suggestion
In this approach, we will implement something similar to Debezium server/engine, but using tools and libraries such as [wal2json](https://github.com/eulerto/wal2json) and [wal-g](https://github.com/wal-g/wal-g).
```

The steps in this approach would be:

> 1. Using Wal-g, we create a base backup of the source database and restore it to the target database. From this point, the target database must be in read-only mode for users other than our plugin user.

Reviewer: ```suggestion
1. Using wal-g, we create a base backup of the source database and restore it to the target database. From this point, the target database must be in read-only mode for users other than our plugin user.
```

2. Using wal2json, we create a stream of transactions from the source database from the point that the base backup was created.
3. A consumer reads the stream and applies the transactions to the target database.

> 4. From some point, we determine that the target database is somehow in sync with the source database, by somehow, I mean that there are less than a threshold number of transactions in the queue. At this point, we can say that databases are 'ready' for migration.

Reviewer: ```suggestion
4. At one point, we try to determine whether the target database is in sync with the source database or not, i.e. there are less than a threshold number of transactions in the queue. At this point, we can say that databases are 'ready' for migration.
```

> 5. When databases are 'ready', and the user confirms, we need to put the source database in read-only mode and the consumer should process the remaining transactions until nothing is remaining.

Reviewer: ```suggestion
5. When databases are 'ready', and the user confirms, we need to put the source database in read-only mode and the consumers should process the remaining transactions until all of them are processed.
```

>    During the time that the consumer is applying the remaining transaction, and the source DB is in read-only mode, there might be some write queries coming to the database. I can think of two ways to handle these queries:

Reviewer: ```suggestion
   While the consumers are applying the remaining transactions, and the source DB is in read-only mode, and there might be some write queries coming to the database. I can think of two ways to handle these queries:
```
